wire xmltools gen targets into our cmake build system

cyrush opened the pull request:
### Description

wire xmltools gen targets into our cmake build system
allow easier wielding of our xml tools to regen sources
document cases where we can't use code gen

### Type of change

Please select one below (*Please click check boxes AFTER submitting ticket*)

- [ ] Bug fix
- [x] New feature
- [ ] New Documentation
- [ ] Other (please describe below)

cyrush committed:
cmake: add code gen helpers, wire up core python and java code gen srcs

cyrush committed:
rename xml tool sources to match purpose (no more main{1-4,6-10})

cyrush committed:
cmake gen targets for plots, updates to xml2cmake

cyrush committed:
cmake add code gen helpers to operators

cyrush committed:
cmake add code gen targets for dbs (except for Uintah and Moab, which are custom)

cyrush committed:
docs: very brief dev docs on xmltools, note about cmake intergration

cyrush commented:
This is a WIP

cyrush committed:
allow moab to autogen cmakelists.txt

cyrush committed:
add rest of xml2atts code gen targets

cyrush committed:
add more to xmltools docs section on cmake targets

cyrush committed:
quite config time msgs about code gen tgts, add verbose option, more docs changes about xmltools cmake integration

cyrush commented:
This is ready for review.

cyrush committed:
defer cmake lang init for java case

biagas commented on src/avt/DBAtts/CMakeLists.txt:
```diff
@@ -4,9 +4,41 @@
+# Add XML Code Gen Targets
+
+set(METADATA_GEN_OBJECTS 
+    avtVectorMetaData
```
I suggest alphabetizing this list.

cyrush replied:
good idea, I will do this!

biagas commented on src/common/CMakeLists.txt:
```diff
@@ -40,8 +40,132 @@
+# Add XML Code Gen Targets
+
+set(COMMON_STATE_GEN_OBJECTS 
+    ProcessAttributes
```
I suggest alphabetizing this list.

biagas commented on src/databases/MOAB/CMakeLists.txt:
```diff
@@ -53,31 +55,33 @@ ${VISIT_INCLUDE_DIR}/avt/Pipeline/Sources
 )
 
-LINK_DIRECTORIES(${VISIT_LIBRARY_DIR} ${MOAB_LIBRARY_DIR} ${HDF5_LIBRARY_DIR})
+LINK_DIRECTORIES(${VISIT_LIBRARY_DIR} )
```
Do these changes to the link_directories and target_link_libraries work for the MOAB reader?  I thought there was some funniness with this plugin that required custom CMakeLists.txt?

cyrush replied:
this was a mistake, thanks for catching .

biagas commented on src/doc/dev_manual/xmltools.rst:
```diff
@@ -0,0 +1,78 @@
+CMake Integration
+--------------------
+
+We rely on xml code generation to keep our State object, Attribute, and Plugin APIs up-to-date. To automate the process we provide CMake targets that call our xml code generation tools for each object or plugin registred. Individual code gen targets are all wired into top level targets that allow you to apply
```
I thought we had agreed to limit line length in our .rst files.  If I am incorrect, please disregard this comment. Otherwise, please adjust line length appropriately.

cyrush replied:
I don't recall, but happy to do so -- what did we agree on 80 chars?

markcmiller86 replied:
My recollection is that we set an 80 column limit at the first doc hack-a-thon we had in transision from `.doc` to `.rst`. IMHO, 80 column is some kind of a hold-over from the 1970's when we had only 80 column CRT terminals. A larger limit would be better. And, I think there are many places where 80 is not followed now either because someone forgot or because some markdown constructs cannot tolerate line breaks. I wouldn't sweat the line limit here as long as there are indeed lines (in another project I work on a key VIP likes to document a single line to a whole paragraph which simply thwarts standard text diffing tools) and is just a pain to work with.

cyrush replied:
thanks @markcmiller86 -- I did change to 80 col, in this case -- I really i think I needed more lines as well (it was a big long line with multiple sentences)

biagas commented on src/plots/FilledBoundary/FilledBoundaryPluginInfo.h:
```diff
@@ -3,7 +3,7 @@
 
 // ************************************************************************* //
-//                     FilledBoundaryPluginInfo.h                            //
+//                               FilledBoundaryPluginInfo.h                            //
```
I wonder if it would be better if our code gen for this would output:
<pre>
// **********************************************************
//  FilledBoundaryPluginInfo.h
// **********************************************************
</pre>

instead of trying to 'center' the plugin name and having our end // not align ?
Since you are regenerating all the info objects, it would be a good time to make the change.

cyrush replied:
I agree, can we add a new ticket for this change?

biagas replied:
Done #4177

markcmiller86 replied:
Ok, forget my previous comment on this issue in another file. I like the new idea of NOT centering these.

cyrush replied:
thanks for adding #4177

biagas commented on src/sim/V2/lib/CMakeLists.txt:
```diff
@@ -19,13 +19,37 @@
+# Add XML Code Gen Targets
+
+set(LIBSIM_GEN_OBJECTS 
+    MessageMetaData
```
I suggest alphabetizing this list.

markcmiller86 commented on src/CMake/PluginMacros.cmake:
```diff
@@ -85,3 +85,107 @@ MACRO(VISIT_PLUGIN_TARGET_FOLDER type pname)
     SET_TARGET_PROPERTIES(${ARGN} PROPERTIES FOLDER "plugins/${type}/${pname}")
 ENDMACRO(VISIT_PLUGIN_TARGET_FOLDER)
 
+
```
Does this really create 6 make targets per plugin? My recollection from working in the data directory is that large numbers of targets causes CMake to take forever to complete. It already (now) can take more than a minute on my OSX system. Just curious, have you looked at performance of CMake'ing VisIt with all these new targets defined?

cyrush replied:
to answer: yes it does create 6 targets per plot and operator plugin and 2 per database plugin

markcmiller86 commented on src/CMake/VisItMacros.cmake:
```diff
@@ -120,3 +120,196 @@ MACRO(VISIT_PLUGIN_TARGET_RTOD type)
 ENDMACRO(VISIT_PLUGIN_TARGET_RTOD)
 
 
+##############################################################################
```
These 5 functions have about 95% of code in common. Can't they be refactored maybe to one function with some additional params or macro'ized?

cyrush replied:
we could, but since I have already tested this I would rather focus on my time on python changes and not stay stuck in cmake land for too much longer

markcmiller86 replied:
Do you think re-factoring to a single func. would reduce maint. burden? If so, I think its worth doing. As things stand now, if something needs to be adjusted in the way these targets work, we have 5 times as much editing and reviewing to do.

cyrush replied:
yes - done the right way it could. Also given we need to special case for windows that tips me over to support the changes.  Let's address as part of another PR when we can attack the windows support as well or?

markcmiller86 replied:
Ok, will assume you will add a ticket, ok?

cyrush replied:
I went ahead and tackled it, and also added the case for windows.

markcmiller86 replied:
❤️

markcmiller86 commented on src/CMakeLists.txt:
```diff
@@ -741,6 +741,7 @@ IF(WIN32)
   ENDIF(VISIT_WINDOWS_APPLICATION)
 ENDIF(WIN32)
 
+OPTION(VISIT_CMAKE_VERBOSE_GEN_TARGET_MESSAGES "Display names and details about code gen targets during CMake configure time." OFF)
```
This seems to make *messages* during code-gen targets enabled or not. Following up on CMake performance...one way to address it if it is a problem is to also have a variable that conditionally defines code-gen targets only for *dev* builds. That way, users and maintainers would be suffering the possible performance hit

cyrush replied:
It takes quite a long time to configure and generate visit using cmake already. 

I didn't measure how much more time this adds.  I guess we could add an option to completely turn off creating these targets and measure.

My observation from this exercise:
We don't have good hygiene with respect to applying all of our code gen tools consistently. 
These CMake targets can help a ton with this, I think its worth it to have them on.

markcmiller86 replied:
I didn't mean to sugges we don't do this...only that we should maybe consider additional controls w.r.t. to when these targets get created.

markcmiller86 commented on src/databases/ADIOS/CMakeLists.txt:
```diff
@@ -61,6 +63,9 @@ ${VISIT_INCLUDE_DIR}/avt/Pipeline/Sources
 ${VISIT_INCLUDE_DIR}/avt/VisWindow/VisWindow
 ${VISIT_INCLUDE_DIR}/visit_vtk/full
 ${VISIT_INCLUDE_DIR}/visit_vtk/lightweight
+${VTKh_INCLUDE_DIRS}
```
This change involving vtkm-1.2 dependencies seems unrelated to code-gen targets.

cyrush replied:
That comes from running our current code gen tools -- it's a sign that our CMakeLists are out of date with respect to what our current code gen tools do. 

I agree we should think about this, but I think it's orthogonal  to wiring up these tools.

markcmiller86 replied:
If you believe there is an orthogonal issue here, would a new issue ticket be in order?

It just seems to me that we don't often visit all these files. So, when we do (both you as the author in this case and others of us as reviewers), it might make sense to leverage that time just a bit more and fix the issues that that no longer make sense instead of kicking them down the road. 

At the same time, I understand expediency and that you don't think anything here makes things worse than they already were.

markcmiller86 commented on src/databases/MFEM/CMakeLists.txt:
```diff
@@ -3,6 +3,8 @@ PROJECT(MFEM_database)
 
 INCLUDE(${VISIT_SOURCE_DIR}/CMake/PluginMacros.cmake)
 
+ADD_DATABASE_CODE_GEN_TARGETS(MFEM)
```
No vtk-h related dependencies here?

cyrush replied:
again - this is just the result of our current xml2cmake. 

Pretty sure, plots and operators need to know about vtk-h and vtk-m, but not databases.

biagas replied:
Only those plots and operators actually utilizing vtk-h need the includes, but they were added to the code gen to make it easier for development of future plugins (?). IMHO, I think they should be added to the xml files of those plugins requiring them instead of applied to all plugins via code gen of CMakeLists.txt

cyrush replied:
I agree, we can add a ticket to follow up.

markcmiller86 replied:
> Only those plots and operators actually utilizing vtk-h need the includes, but they were added to the code gen to make it easier for development of future plugins (?). IMHO, I think they should be added to the xml files of those plugins requiring them instead of applied to all plugins via code gen of CMakeLists.txt

Perhaps the right answer here is to recognize when code-gen tools are being used to *jump-start* development of a new plugin vs. re-gening existing code. I think it makes sense for the *jump-start* mode to possibly include (or via comments suggest including) commonly needed header files, etc. But, that service is really to help newbie/novice developers overcome common issues and is not something we want to impact *mature* code. So, maybe we should have an additional CL switch to the code-gen tools that disables this from within these targets but which is otherwise on by default. That would address the reasons this boilerplate code is in the code-gen outputs to begin with but would prevent it from polluting re-gens of mature code. Maybe a new issue ticket?

cyrush replied:
Note I added: #4179

cyrush replied:
My specific observation is: there isn't a compelling case to have them at all, given they are core-ish dependencies -- they should be included by indirection.

markcmiller86 commented on src/databases/Uintah/CMakeLists.txt:
```diff
@@ -1,7 +1,15 @@
 
 INCLUDE(${VISIT_SOURCE_DIR}/CMake/PluginMacros.cmake)
+#
+# NOTE: THIS FILE HAS HAND MODIFICATIONS !!! 
```
Since you are already visiting all these cases, why not just add solution for #4106 and use that here as opposed to these comments?

cyrush replied:
I just wanted to try to capture what could not be wired up during this pass.

For the two cases we have for database plugins that can't be gen'ed they are actually minor.  My suggestion would be to not maintain manual edits at all for these cases, but to rework to make sure they could be code-gened.

markcmiller86 commented on src/doc/dev_manual/index.rst:
```diff
@@ -11,6 +11,7 @@ VisIt Developer Manual
     pr_review
     RCDevelopment
     test_suite
+    xmltools
```
Awesome! 👍

markcmiller86 commented on src/operators/Remap/CMakeLists.txt:
```diff
@@ -3,6 +3,8 @@ PROJECT(Remap_operator)
 
 INCLUDE(${VISIT_SOURCE_DIR}/CMake/PluginMacros.cmake)
 
+ADD_OPERATOR_CODE_GEN_TARGETS(Remap)
```
No vtkm dependencies here?

markcmiller86 commented on src/operators/Slice/CMakeLists.txt:
```diff
@@ -3,6 +3,8 @@ PROJECT(Slice_operator)
 
 INCLUDE(${VISIT_SOURCE_DIR}/CMake/PluginMacros.cmake)
 
+ADD_OPERATOR_CODE_GEN_TARGETS(Slice)
```
No vtkm dependencies here?

markcmiller86 commented on src/plots/Boundary/BoundaryPluginInfo.h:
```diff
@@ -3,7 +3,7 @@
 
 // ************************************************************************* //
-//                           BoundaryPluginInfo.h                            //
+//                               BoundaryPluginInfo.h                            //
```
Uneeded white space change?

markcmiller86 commented on src/tools/dev/xml/CMakeLists.txt:
```diff
@@ -10,26 +10,11 @@
 #   Kathleen Bonnell, Mon Jan 10 07:59:28 MST 2010
 #   Remove PROGRAM9 (xml2projectfile), and PROGRAM5 (xml2makefile).
 #
+#   Cyrus Harrison, Mon Dec  9 13:39:19 PST 2019
```
:heart:

markcmiller86 reviewed (changes requested):
I have asked several questions. Key among these...why does almost every plugin (plot, database or operator) seem to have dependence on `winutil, visit_vtk/full, visit_vtk/lightweight` and now with this PR, adding `VTKh_INCLUDE_DIRS, vtkm-1.2, vtkm-1.2/vtkm/thirdparty/tauotuple`?

```
{VISIT_INCLUDE_DIR}/winutil
 ${VISIT_INCLUDE_DIR}/visit_vtk/full		 ${VISIT_INCLUDE_DIR}/visit_vtk/full
 ${VISIT_INCLUDE_DIR}/visit_vtk/lightweight		 ${VISIT_INCLUDE_DIR}/visit_vtk/lightweight
 ${VTKh_INCLUDE_DIRS}
 ${VTKM_DIR}/include/vtkm-1.2
 ${VTKM_DIR}/include/vtkm-1.2/vtkm/thirdparty/taotuple
```

Also, It looks like this is just *defining* new targets and hasn't actually employed them to re-gen any of our code-generated stuff, right? If so, ignore another remark I made regarding this when I was thinking about this PR with a different thinking cap.

cyrush committed:
pr feedback: alpha order obj lists and imporve docs

cyrush committed:
re-fix moab

cyrush commented:
for the database plugins -- I needed to re-run the code gen tools to make sure the cmake function that wires up the code-gen tools is correctly inserted post gen. That is why there are lots of changes for the CMakeLists.txt files for our plugins

biagas commented:
### @biagas original comment...

General question. I'm trying out the PR on windows. I was actually surprised that the codegen is ON by default. Is that the design or did I just miss some flag to turn it all off? There are some changes needed to get this working on Windows. I'm working on that now. Also, it would be great if all the codegen targets were in folders (maybe a toplevel 'codegen' folder, then 'codegen_cpp', 'codgen_cmake', etc underneath that). Cleans up the VS interface quite a bit. I will see where this can be added once I figure out the best folder structure for them.

I say this b/c I totally missed the idea bout adding the code gen folder until I reviewed my email.


### @markcmiller86 accidental edit (which was supposed to respond to just the quoted line)

> I was actually surprised that the codegen is ON by default.

So, at a *minimum* I think it should be conditional and *on* by default only for *dev* builds. In other words, off for users/maintainers building in the wild (primarily because of my belief that the additional targets cause CMake to run *much* longer. As @cyrush says, it is long already but I would not be surprised if this tripled CMake-time and if so, its worth making conditional).

If we make wholly conditional with a sep. flag (apart from inferring from build type) then when a developer discovers they need to re-gen, won't they be *required* to re-cmake? Or, do developers still have the option in that case of re-gen'ing manually as we do now?

biagas commented on src/CMake/VisItMacros.cmake:
```diff
@@ -120,3 +120,196 @@ MACRO(VISIT_PLUGIN_TARGET_RTOD type)
+        MESSAGE(STATUS "Adding xml tools C++ generation target: ${gen_target_name}")
+    endif()
+
+    set(xml_gen_tool ${CMAKE_BINARY_DIR}/bin/xml2atts)
```
In order to get this working on Windows, I needed to change this line to:
<pre>   
     if(NOT WIN32)
        set(xml_gen_tool ${CMAKE_BINARY_DIR}/bin/xml2atts)
    else()
        set(xml_gen_tool ${VISIT_EXECUTABLE_DIR}/visit.exe -xml2atts)
    endif()
</pre>
Will need to do similar for the other macros.
But in the short term, I think all of this needs to be turned off on windows because the code-gen tools generate files with windows-style line endings, which marks all the files as changed. I am entering a ticket to see if the tools can be changed to output unix-style line endings always.

cyrush commented:
I can do a quick check for the overhead by timing configure for develop vs configure for this branch to inform if we should be concerned. If it all possible I think having them on by default is the best path. Each option off by default becomes a barrier to use. (pseudo related to my thinking for  https://github.com/visit-dav/visit/issues/4118)

@biagas thanks for the cases on windows that I overlooked. What do you think the best near term path is?

My goal is to be able to use these to help me support the changes for python code gen.

biagas commented:
For the short term, I think disable it completely on Windows since the code generation unnecessarily modifies files by adding windows-style line endings. I can then work separately on fixing the problems on Windows.  

I do think, however, that a flag to turn off this codegen is absolutely necessary.  Custom targets are always considered out-of-date by cmake, so are always re-built with each make.  That's a lot of unnecessary regeneration during a development cycle.  
We might want to consider adding the EXCLUDE_FROM_ALL and/or EXCLUDE_FROM_DEFAULT_BUILD target properties flags.  That way the targets are created, but not built by default.

cyrush reviewed (commented):
@biagas these are not wired into the default target, so they are only executed when the targets are invoked. 

or that was my intent ... maybe I forgot the proper magic as I refactored to create the custom targets -- I'll double check.

cyrush commented:
Ok, I think i did this correctly with respect to not always being on.

I do *not* use the ``ALL``option to `add_custom_target` when creating the targets. 

```
ALL
Indicate that this target should be added to the default build target so that it will be run every time (the command cannot be called ALL).
```

(docs from: https://cmake.org/cmake/help/latest/command/add_custom_target.html)

cyrush committed:
refactor marcos to use common helper (needs more testing)

cyrush commented:
@biagas I added a case for windows when refactoring the macros.


what I think is still pending:

test configure time difference and add option to skip targets

cyrush commented:
reviewing my email and I am a bit confused:
for @biagas ' comment that starts with:
> > I was actually surprised that the codegen is ON by default.

Seems this was edited by mark?

From email I see more text from kathleen:

General question. I'm trying out the PR on windows. I was actually surprised that the codegen is ON by default. Is that the design or did I just miss some flag to turn it all off? There are some changes needed to get this working on Windows. I'm working on that now. Also, it would be great if all the codegen targets were in folders (maybe a toplevel 'codegen' folder, then 'codegen_cpp', 'codgen_cmake', etc underneath that). Cleans up the VS interface quite a bit. I will see where this can be added once I figure out the best folder structure for them.

I say this b/c I totally missed the idea  about adding the codegen folder for VS until I reviewed my email.

markcmiller86 commented:
Ok, Using Cyrus' email here, I went back and adjusted the comment so that this text is again there.

But, to leave things in basically the same order, I also left my edits as a sub-block later in the same comment.

Mark


"Cyrus Harrison" wrote:


reviewing my email and I am a bit confused:
for @biagas<https://github.com/biagas> ' comment that starts with:

I was actually surprised that the codegen is ON by default.
Seems this was edited by mark?

From email I see more text:

General question. I'm trying out the PR on windows. I was actually surprised that the codegen is ON by default. Is that the design or did I just miss some flag to turn it all off? There are some changes needed to get this working on Windows. I'm working on that now. Also, it would be great if all the codegen targets were in folders (maybe a toplevel 'codegen' folder, then 'codegen_cpp', 'codgen_cmake', etc underneath that). Cleans up the VS interface quite a bit. I will see where this can be added once I figure out the best folder structure for them.

I say this b/c I totally missed the idea bout adding the code gen folder until I reviewed my email.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/visit-dav/visit/pull/4172?email_source=notifications&email_token=ABLUUZGP6LAJIBBHXUAGAS3QYHJRTA5CNFSM4JYVT2MKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGVS6LI#issuecomment-564866861>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABLUUZA76YF6GIXCYPSUMRDQYHJRTANCNFSM4JYVT2MA>.

markcmiller86 commented:
> Each option off by default becomes a barrier to use

I reacted with 😕 (confusion) to this statement because I guess I thought I was seeing a general philosophical argument and would like to understand it better. Perahps, at some point when you have more time, you can elaborate.

biagas commented:
> what I think is still pending:
> 
> test configure time difference and add option to skip targets

There is a difference between configuration and generation times.  When I tried out the branch on linux, and issued the cmake command, I noticed more of a slowdown during generation, eg I saw the 'done with configure' comment and the 'done with generation' was a very long time after that.
As an aside, a found a VTK blog that hints CMake 3.11 performs much better than older versions:
https://blog.kitware.com/improving-cmakes-runtime-performance/
We may want to revisit the idea of modernizing our CMake usage. (#1636 comes to  mind)

I did notice, as well, that on linux, the gen_xx targets aren't built automatically like they are on Windows.

cyrush commented:
@biagas yes, I am going to test time to start configure past generation. 

I agree generation seems to take much longer than configure.

Thanks for letting me know that the targets are auto built on windows -- that is not what I intended, we will need to explore why not using `ALL` doesn't work the same way on windows. 

For now, I will disable these targets on windows.

cyrush commented:
@markcmiller86 perhaps we can discuss more at a future meeting?

markcmiller86 commented:
> As an aside, a found a VTK blog that hints CMake 3.11 performs much better than older versions:

If that fixes any potential performance issues then you can ignore all my comments here regarding performance.

In general, it doesn't make sense to me that CMake *generation* takes as long as it does even if we had *thousands* of targets (except maybe on file systems where many tiny files are a problem so either nfs mounts or lustre/gpfs). So, its clear to me that there is indeed a serious issue internal to CMake with generation performance and if they've fixed it, great!! 👍

biagas commented:
Just for information for the future of the code_gen stuff on windows:  If they aren't needed for a given development cycle, then I would like an option to not even create them.  Even after we fix the problem on windows where they are part of the all-build, then they still will pollute the VS IDE and slow down initial loading considerably.  May not seem like much to you, but it honestly is frustrating.  Without parallel targets, jumped from 914 project files (without gen_xx targets)  to 2097 project files with them.

cyrush committed:
address pr feedback

cyrush commented:
@biagas I removed the option for verbose messages and added `VISIT_CREATE_XMLTOOLS_GEN_TARGETS` which controls creation of the targets. 
That should help.

Right now it defaults to ON, but additionally I am not adding targets at all for WIN32, given outstanding issues. 

I also did some timings on my mac:

With code-gen targets on:
```
real	1m24.564s
user	0m57.653s
sys	0m21.719s
```

w/o code-gen targets on:
```
real	1m11.282s
user	0m52.233s
sys	0m16.828s
```

I would expect it to be a bit worse on HPC systems depending on the file system, but I don't think it is a huge cost.  Based on these timings, I would like to leave the default to ON (for macOS and linux)

cyrush committed:
Merge branch 'develop' into task/2019_12_cmake_regen_targets

markcmiller86 commented:
@cyrush thanks for taking a look at performance (👍). Based on those numbers (which are somewhat surprising given my recollections), I agree....there doesn't appear to be any need to make code-gen targets conditional at least on OSX and Linux.

cyrush commented:
since windows cases are shutoff -- this shouldn't (tm) cause any windows issues and we can work to include windows later
